### DISCUSSION TITLE
Should miette be eyreish, or should it be thiserrorish

### DISCUSSION THREAD

**Discussion by CAD97** [zkat/miette#128] [Ideas]
My take: miette is doing a bit too much by being an eyreish and providing a `Report` type. It should focus on just providing the `Error` type (derive).

For use in libraries, you already recommend just using `#[derive(Error, Diagnostic)]`, and not using `Report`.

For use in applications, you provide `Report` and `IntoDiagnostic`. `IntoDiagnostic` is effectively lossy: it wraps an arbitrary `impl Error` into effectively a "null object" `impl Diagnostic` which provides `None` for all of its diagnostic info. The underlying error is seen through `Error::source`.

`GraphicalReportHandler` does render the source chain, at least, and `wrap_err` maintains the diagnostic information.

But I'm not certain exactly what benefit `miette::Report` provides over using `eyre::Report`. It feels to me that what `miette` wants is an eyre [`Section`](https://docs.rs/color-eyre/0.6.1/color_eyre/section/trait.Section.html) for the diagnostic information.

What I'm most concerned about, I guess, is that I make a nice `miette` diagnostic for my library's error, and then none of it is seen unless the application sticks that error into a `miette::Report`. If the error is wrapped by another library, the information is lost unless it passes through the diagnostic info.

I honestly don't know the "correct" or "best" solution here. But at least to me it feels like miette is trying to be two things -- a general purpose error reporting library, and a specialized parsing error library -- but not really _excelling_ at either. (Specifically, the requirement of `'static` gets in the way of use for parsing libraries which parse `&str`, since it requires a source clone.)

What I'm doing for the right now in my library is just providing a non-`'static` `Diagnostic` and pointing at how to use a `ReportHandler` to render it, providing a `static_diagnostic() -> impl Copy + Diagnostic` which strips the source information (since it's a parsing library, the caller has the source and can re-inject it), and a `to_owned() -> miette::Report` which uses `with_source_code` on `miette!(self.static_diagnostic())` to attach the source info in an owned manner.

At the very least, though, even if `Report` is kept, a decent amount of eyre's complexity could be trimmed off; the fact that the reporting hook creates a separate report handler for every `miette::Report` seems excessive; I don't really see a use case for downcasting and providing different report handlers for different concrete diagnostic types, and just registering a global `&'static ReportHandler` can reduce complexity there.

I'll be completely honest and say that as part of writing this down I'm less certain in my position now, but it still _feels_ to me that miette can and should use eyre rather than being an eyreish. (The point of eyre is to be pluggable, after all, _especially_ color-eyre.)

**Comment by CAD97** [comment]
To be completely transparent:

The reason this has come up for me is I'm using miette to report parsing errors and building a serde deserializer on top of that. The chance of someone using miette to `Report` errors from a serde deserialization is approximately zero, but I would still like to provide nice errors. I'm still figuring out how best to do so. Having a `get_hook()` that can handle non-`'static` reports would allow me to stick the formatting into `Debug` and/or `Display`.

**Comment by zkat** [comment]
eyreish has always been a temporary measure, until [provide-any](https://github.com/nrc/rfcs/blob/dyno/text/0000-dyno.md) landed, so you could have `Diagnostic`s without being lossy. I guess I don't actually understand how a Section could be used and still have access to `Diagnostic` fields?

**Reply by CAD97** [reply]
I'll be transparent: I thought that (color-)eyre provided a pluggable solution for provide-any, and that's how spantraces are attached. But instead, it's just all added to color-eyre's `Handler`. (The advantage would be getting #14 implicitly, of course.)

I guess what I think I want is for `MyDiagnostic as Display` to display the global-hook-ily formatted diagnostic. Then miette diagnostics would be trivially pluggable into an anyhow/eyre error stack.

(And for color-eyre, what I think would be _ideal_, even if not realistic, would be for the current `Display` message to be provided in the error chain, and the rest of the diagnostic attatched in a `━━━━━━━━ DIAGNOSTIC ━━━━━━━━` section. IIUC after further study of eyre/color-eyre, though, this would require _another_ layer of hooking to allow miette to attach a `.with_section` when converting `MyDiagnostic` into `eyre::Report (with color_eyre::Handler)`, as well as the section callback having some way to access `MyDiagnostic`, so... complicated.)

TL;DR: There's more issues with doing it (what I think/thought to be) the "proper" way than I initially realized (duh!) and the current solution is a fine one for the time being, until dyno/provide-any are usable.
